res/res_pjsip_device_features.c

```diff
@@ -0,0 +1,1005 @@
+		feature_state_sub->sync_data.fwd_exten_noanswer, sizeof(feature_state_sub->sync_data.fwd_exten_noanswer))) {
+		feature_state_sub->have_callforwardnoanswer = 1;
+	}
+	if (!ast_db_get("subscription_device_features/ringcount", endpoint_name, tmp, sizeof(tmp))) {
+		feature_state_sub->sync_data.ring_count = atoi(tmp);
+	}
+
+	/* 2.1.1: When the phone first SUBSCRIBEs, the body is empty, and thus it needs an update on everything. */
+	update_everything(feature_state_sub, &feature_state_sub->sync_data);
+
+	/* Insert into the linked list. */
+	subitem->sub = feature_state_sub;
+	AST_RWLIST_WRLOCK(&sublist);
+	AST_RWLIST_INSERT_HEAD(&sublist, subitem, entry);
+	AST_RWLIST_UNLOCK(&sublist);
+
+	return feature_state_sub;
+}
+
+struct notify_task_data {
+	struct ast_sip_device_feature_sync_data sync_data;
+	struct feature_state_subscription *feature_state_sub;
+	int terminate;
+};
+
+static void notify_task_data_destructor(void *obj)
+{
+	struct notify_task_data *task_data = obj;
+	ao2_ref(task_data->feature_state_sub, -1);
+}
+
+static void dump_sync_data(struct ast_sip_device_feature_sync_data *sync_data)
+{
+	ast_debug(3, "DND update needed: %d\n", sync_data->update_needed_dnd);
+	ast_debug(3, "Forward Always update needed: %d\n", sync_data->update_needed_fwd_always);
+	ast_debug(3, "Forward Busy update needed: %d\n", sync_data->update_needed_fwd_busy);
+	ast_debug(3, "Forward No Answer update needed: %d\n", sync_data->update_needed_fwd_noanswer);
+}
+
+static struct notify_task_data *alloc_notify_task_data(struct feature_state_subscription *feature_state_sub)
+{
+	struct notify_task_data *task_data = ao2_alloc(sizeof(*task_data), notify_task_data_destructor);
+
+	if (!task_data) {
+		ast_log(LOG_WARNING, "Unable to create notify task data\n");
+		return NULL;
+	}
+
+	task_data->feature_state_sub = feature_state_sub;
+	ao2_ref(task_data->feature_state_sub, +1);
+	ast_debug(2, "Allocating notify task\n");
+
+	return task_data;
+}
+
+static int notify_task(void *obj)
+{
+	RAII_VAR(struct notify_task_data *, task_data, obj, ao2_cleanup);
+	struct ast_sip_body_data data = {
+		.body_type = AST_SIP_DEVICE_FEATURE_SYNC_DATA,
+		.body_data = &task_data->sync_data,
+	};
+
+	/* The subscription was terminated while notify_task was in queue.
+	 * Terminated subscriptions are no longer associated with a valid tree, and sending
+	 * NOTIFY messages on a subscription which has already been terminated won't work. */
+	if (ast_sip_subscription_is_terminated(task_data->feature_state_sub->sip_sub)) {
+		return 0;
+	}
+
+	/* Pool allocation has to happen here so that we allocate within a PJLIB thread */
+	ast_sip_subscription_notify(task_data->feature_state_sub->sip_sub, &data, task_data->terminate);
+	return 0;
+}
+
+#define NUM_FEATURES 4
+static int send_update(struct feature_state_subscription *feature_state_sub)
+{
+	int i;
+
+	/* XXX This is kind of a kludge. In order to send updates for more than one feature,
```

Comment:
Why do this "kind of a kludge" and send many NOTIFYs instead of just one with multipart?
Please look at pjsip_multipart_* functions.
https://www.pjsip.org/pjsip/docs/html/group__PJSIP__MULTIPART.htm

Reply:
According to https://www.cisco.com/c/dam/en/us/td/docs/voice_ip_comm/broadworks/FD/AS/DeviceFeatureKeySynchronizationFD.pdf
2.2 Call Flows
2.2.1 Initial SUBSCRIBE and NOTIFY
"When a supported phone powers up, it sends a SUBSCRIBE request to the Application
Server to create a subscription for the as-feature-event package and to get the **full** feature
status. "

"The Application Server sends a NOTIFY request to the phone
with a body that contains the **complete** feature status, and the phone sends a 200
response. The body is a **MIME multipart body**, containing an XML document for **each**
feature assigned to the subscriber".